casr/src/triage.rs

```diff
@@ -0,0 +1,444 @@
+            info!("Using {} threads", num_of_threads - 1);
+            let counter = RwLock::new(0_usize);
+            let total = crashes.len();
+            let custom_pool = rayon::ThreadPoolBuilder::new()
+                .num_threads(num_of_threads)
+                .build()
+                .unwrap();
+            let at_index = gdb_args
+                .iter()
+                .skip(1)
+                .position(|s| s.contains("@@"))
+                .map(|x| x + 1);
+            custom_pool
+                .join(
+                    || {
+                        crashes.par_iter().try_for_each(|crash| {
+                            if let Err(e) = (CrashInfo {
+                                path: crash.to_path_buf(),
+                                target_args: gdb_args.clone(),
+                                at_index,
+                                casr_tool: casr_gdb.clone(),
+                            })
+                            .run_casr(None, timeout)
+                            {
+                                // Disable util::log_progress
+                                *counter.write().unwrap() = total;
+                                bail!(e);
+                            };
+                            *counter.write().unwrap() += 1;
+                            Ok::<(), anyhow::Error>(())
+                        })
+                    },
+                    || log_progress(&counter, total),
+                )
+                .0?;
+        }
+    }
+
+    let casr_cli = get_path("casr-cli")?;
+    // Print summary
+    let status = Command::new(casr_cli)
+        .arg(dir)
+        .stderr(std::process::Stdio::inherit())
+        .stdout(std::io::stderr().as_fd().try_clone_to_owned()?)
+        .status()
+        .with_context(|| "Couldn't launch casr-cli".to_string())?;
+
+    if !status.success() {
+        error!("casr-cli exited with status {status}");
+    }
+
+    // Report ooms.
+    let oom_dir = dir.join("oom");
+    let oom_cnt = fs::read_dir(&oom_dir).unwrap().count();
+    if oom_cnt != 0 {
+        info!(
+            "{} out of memory seeds are saved to {:?}",
+            oom_cnt, &oom_dir
+        );
+    } else {
+        fs::remove_dir_all(&oom_dir)?;
+    }
+
+    // Report timeouts.
+    let timeout_dir = dir.join("timeout");
+    let timeout_cnt = fs::read_dir(&timeout_dir).unwrap().count();
+    if timeout_cnt != 0 {
+        info!(
+            "{} timeout seeds are saved to {:?}",
+            timeout_cnt, &timeout_dir
+        );
+    } else {
+        fs::remove_dir_all(&timeout_dir)?;
+    }
+
+    // Check bad reports.
+    if let Ok(err_dir) = fs::read_dir(dir.join("clerr")) {
+        warn!(
+            "{} corrupted reports are saved to {:?}",
+            err_dir
+                .filter_map(|x| x.ok())
+                .filter_map(|x| x.file_name().into_string().ok())
+                .filter(|x| x.ends_with("casrep"))
```

Comment:
check extension

Reply:
Also, there may be non-corrupted generated gdb reports. We should not count them twice.

Reply:
crash-abc
crash-abc.casrep
crash-abc.gdb.casrep

count = 1